File: src/main/java/com/github/bzumhagen/sct/ChangelogConfiguration.java
package com.github.bzumhagen.sct;

import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;

import java.util.HashSet;
import java.util.Set;
import java.util.regex.Pattern;

final class ChangelogConfiguration {
    private Pattern versionPattern;
    private Pattern tagPattern;
    private Pattern referencePattern;
    private Set<String> tags;

    ChangelogConfiguration() {
        Config conf = ConfigFactory.load();
        versionPattern = Pattern.compile(conf.getString("sct.patterns.version"));
        tagPattern = Pattern.compile(conf.getString("sct.patterns.tag"));
        referencePattern = Pattern.compile(conf.getString("sct.patterns.reference"));
        tags = new HashSet<>(conf.getStringList("sct.tags"));
    }

    Set<String> getTags() {
        return tags;
    }

    Pattern getVersionPattern() {
        return versionPattern;
    }

    Pattern getTagPattern() {
        return tagPattern;
    }

    Pattern getReferencePattern() {
        return referencePattern;
    }
}

File: src/main/scala/com/github/bzumhagen/sct/ChangelogConfiguration.scala
package com.github.bzumhagen.sct

import com.typesafe.config.{Config, ConfigFactory}

import scala.collection.JavaConverters._
import scala.util.matching.Regex

object ChangelogConfiguration {
  def load(config: Config = ConfigFactory.load()): ChangelogConfiguration = {
    new ChangelogConfiguration(
      name = config.getString("sct.name"),
      versionPattern = config.getString("sct.patterns.version").r,
      tagPattern = config.getString("sct.patterns.tag").r,
      referencePattern = config.getString("sct.patterns.reference").r,
      tags = config.getStringList("sct.tags").asScala.toSet,
      smartGrouping = config.getBoolean("sct.smartGrouping"),
      showReference = config.getBoolean("sct.showReference")
    )
  }
}

case class ChangelogConfiguration(
  name: String,
  versionPattern: Regex,
  tagPattern: Regex,
  referencePattern: Regex,
  tags: Set[String],
  smartGrouping: Boolean,
  showReference: Boolean
)
